numcosmo/math/ncm_stats_dist_nd_kde_studentt.c

```diff
@@ -0,0 +1,1266 @@
+  ncm_matrix_clear (&self->log_cov);
+  ncm_matrix_clear (&self->sample_matrix);
+  ncm_vector_clear (&self->v);
+
+  self->sample     = ncm_stats_vec_new (dim, NCM_STATS_VEC_COV, TRUE);
+  self->cov_decomp = ncm_matrix_new (dim, dim);
+  self->log_cov    = ncm_matrix_new (dim, dim);
+  self->v          = ncm_vector_new (dim);
+}
+
+static void 
+_ncm_stats_dist_nd_kde_studentt_prepare_cov (NcmStatsDistNdKDEStudenttPrivate * const self)
+{
+  gint i, ret;
+
+  if (self->sample_matrix == NULL)
+  {
+    self->sample_matrix = ncm_matrix_new (self->n, self->d);
+    g_ptr_array_set_size (self->smatrix_rows, 0);    
+    for (i = 0; i < self->n; i++)
+    {
+      NcmVector *row_i = ncm_matrix_get_row (self->sample_matrix, i);
+      g_ptr_array_add (self->smatrix_rows, row_i);
+    }
+  }
+  else if ((self->n != ncm_matrix_nrows (self->sample_matrix)) || (self->d != ncm_matrix_ncols (self->sample_matrix)))
+  {
+    ncm_matrix_clear (&self->sample_matrix);
+    self->sample_matrix = ncm_matrix_new (self->n, self->d);
+
+    g_ptr_array_set_size (self->smatrix_rows, 0);    
+    for (i = 0; i < self->n; i++)
+    {
+      NcmVector *row_i = ncm_matrix_get_row (self->sample_matrix, i);
+      g_ptr_array_add (self->smatrix_rows, row_i);
+    }
+  }
+
+  for (i = 0; i < self->n; i++)
+    ncm_matrix_set_row (self->sample_matrix, i, ncm_stats_vec_peek_row (self->sample, i));
+
+  ret = gsl_blas_dtrsm (CblasRight, CblasUpper, CblasNoTrans, CblasNonUnit, 1.0, ncm_matrix_gsl (self->cov_decomp), ncm_matrix_gsl (self->sample_matrix));
+  NCM_TEST_GSL_RESULT ("_ncm_stats_dist_nd_kde_studentt_prepare", ret);  
+}
+
+static void 
+_ncm_stats_dist_nd_kde_studentt_prepare (NcmStatsDistNd *dnd)
+{
+  NcmStatsDistNdKDEStudentt *dndg = NCM_STATS_DIST_ND_KDE_STUDENTT (dnd);
+  NcmStatsDistNdKDEStudenttPrivate * const self = dndg->priv;
+  gint i;
+
+  self->n = ncm_stats_vec_nitens (self->sample);
+  self->d = ncm_stats_vec_len (self->sample);
+
+  g_assert_cmpuint (self->n, >, 1);
+  
+  ncm_matrix_memcpy (self->cov_decomp, ncm_stats_vec_peek_cov_matrix (self->sample, 0));
+  if (ncm_matrix_cholesky_decomp (self->cov_decomp, 'U') != 0)
+  {
+    if (ncm_matrix_nearPD (self->cov_decomp, 'U', TRUE, self->nearPD_maxiter) != 0)
+    {
+      ncm_matrix_set_zero (self->cov_decomp);
+      for (i = 0; i < self->d; i++)
+      {
+        ncm_matrix_set (self->cov_decomp, i, i, ncm_stats_vec_get_cov (self->sample, i, i));
+      }
+      /*ncm_matrix_log_vals (self->cov_decomp, "# COV: ", "% 22.15g");*/
+      g_assert_cmpint (ncm_matrix_cholesky_decomp (self->cov_decomp, 'U'), ==, 0);
+    }
+  }
+
+  self->href      = self->over_smooth * pow (4.0 / (self->n * (self->d + 2.0)), 1.0 / (self->d + 4.0));
+  self->href2     = self->href * self->href;
+  self->us_lnnorm = log(gsl_sf_gamma((self->st_freedom + self->d) / 2)) - log(gsl_sf_gamma(self->st_freedom / 2)) - (self->d / 2) * (log(M_PI) + log(self->st_freedom)) -(1/2) * ncm_matrix_cholesky_lndet (self->cov_decomp);
```

Comment:
Break down the computation in several steps, this improves readability and do not impact on performance. Remember to always use gdoubles (e.g., use 2.0 instead of 2).

Reply:
Use lgamma instead of log (gsl_sf_gamma (...)), please pay attention to the codding conventions, spaces between function name and parenthesis, space around operators, etc.